src/lxd.py

```diff
@@ -0,0 +1,388 @@
+
+        Args:
+            timeout: Timeout for starting the LXD instance.
+            force: Whether to force start the LXD instance.
+            wait: Whether to wait until the LXD instance started before returning.
+
+        Raises:
+            LxdException: Unable to start the LXD instance.
+        """
+        try:
+            self._pylxd_instance.start(timeout, force, wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            logger.exception("Failed to start LXD instance")
+            raise LxdError(f"Unable to start LXD instance {self.name}") from err
+
+    def stop(self, timeout: int = 30, force: bool = True, wait: bool = False) -> None:
+        """Stop the LXD instance.
+
+        Args:
+            timeout: Timeout for stopping the LXD instance.
+            force: Whether to force stop the LXD instance.
+            wait: Whether to wait until the LXD instance stopped before returning.
+
+        Raises:
+            LxdException: Unable to stop the LXD instance.
+        """
+        try:
+            self._pylxd_instance.stop(timeout, force, wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            logger.exception("Failed to stop LXD instance")
+            raise LxdError(f"Unable to stop LXD instance {self.name}") from err
+
+    def delete(self, wait: bool = False) -> None:
+        """Delete the LXD instance.
+
+        Args:
+            wait: Whether to wait until the LXD instance stopped before returning.
+
+        Raises:
+            LxdException: Unable to delete the LXD instance.
+        """
+        try:
+            self._pylxd_instance.delete(wait)
+        except pylxd.exceptions.LXDAPIException as err:
+            logger.exception("Failed to delete LXD instance")
+            raise LxdError(f"Unable to delete LXD instance {self.name}") from err
+
+    def execute(self, cmd: list[str], cwd: Optional[str] = None) -> Tuple[int, IO, IO]:
+        """Execute a command within the LXD instance.
+
+        Exceptions are not raise if command execution failed. Caller should check the exit code and
+        stderr for failures.
+
+        Args:
+            cmd: Commands to be executed.
+            cwd: Working directory to execute the commands.
+
+        Returns:
+            Tuple containing the exit code, stdout, stderr.
+        """
+        lxc_cmd = ["/snap/bin/lxc", "exec", self.name]
+        if cwd:
+            lxc_cmd += ["--cwd", cwd]
+
+        lxc_cmd += ["--"] + cmd
+
+        result = secure_run_subprocess(lxc_cmd)
+        return (result.returncode, io.BytesIO(result.stdout), io.BytesIO(result.stderr))
+
+
+class LxdInstanceManager:
+    """LXD instance manager."""
+
+    def __init__(self, pylxd_client: pylxd.Client):
+        """Construct the LXD instance manager.
+
+        Args:
+            pylxd_client: Instance of pylxd.Client.
+        """
+        self._pylxd_client = pylxd_client
+
+    def all(self) -> list[LxdInstance]:
```

Comment:
At least from my perspective and in my humble opinion, usually a list or an iterator is returned. If the user of the class wants to use  a tuple, it can be converted.

By the way the pylxd lib uses a list ([here](https://github.com/lxc/pylxd/blob/ee8fe4a458bcd1636df0eac563f070dc9e371550/pylxd/models/instance.py#L294))

Reply:
This interface class might be redesign, so I have put this down as something to take notes of.

Reply:
Minor comments